syz-manager/filter.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"os"
+	"regexp"
+
+	"github.com/google/syzkaller/pkg/cover"
+	"github.com/google/syzkaller/pkg/log"
+)
+
+type CoverFilter struct {
+	enableFilter    bool
+	kcovFilterStart uint32
+	kcovFilterSize  uint32
+	kcovFilterEnd   uint32
+	weightedPCs     map[uint32]float32
+
+	pcsBitmapPath      string
+	targetLittleEndian bool
+	targetArch         string
+}
+
+func (mgr *Manager) initKcovFilter() {
+	covFilterConfig := make(map[string][]string)
+	if err := json.Unmarshal(mgr.cfg.CovFilter, &covFilterConfig); err != nil {
+		log.Logf(0, "no coverage filter is enabled")
+		mgr.kcovFilter.enableFilter = false
+		return
+	}
+	files := covFilterConfig["files"]
+	funcs := covFilterConfig["functions"]
+	rawPCs := covFilterConfig["pcs"]
+	if len(files) == 0 && len(funcs) == 0 && len(rawPCs) == 0 {
```

Comment:
add more log will be good to check.
log.Logf(0, "coverage filter is enabled but no filter given")

Reply:
The code of checking if coverage filter is invalid is in the `createBitmap`.
Such as `"enable_syscalls": []` would not be warned that no syscall was enabled, should an empty filter be warned?
Also, in `initWeightedPCs`, I print enabled functions and files for checking.
Is it proper that using loglevel 0 there?

Reply:
```
Is it proper that using loglevel 0 there?
```
You are right, better to use low level log.

```
Such as "enable_syscalls": [] would not be warned that no syscall was enabled, should an empty filter be warned?
```
I mean the filter may be set as the wrong value, and we may need check it just in case. For example: I set pcs as 0xdeadbeaf, but there are no actual pc for this pc in vmlinux.

Reply:
If we add code to check if any `pc.Frame` match to a coverage filter.
The log of enabled functions and enabled files would be redundant.
So remove it also, right?

Reply:
I think the log of enabled functions and enabled files is useful. 
Their number is also useful.
Maybe change the log level?

Just an advice. Thanks!

Reply:
I am trying my best to understand what you advised.
1. Do you mean that an empty coverage filter [should be warned](https://github.com/google/syzkaller/pull/2017#discussion_r515800493)?
2. Add code to check if all coverage filters are [valid](https://github.com/google/syzkaller/pull/2017#discussion_r516447943)?
3. Except checking if the filters are valid, enabled functions should be printed out [also](https://github.com/google/syzkaller/pull/2017#discussion_r516462371)? And both of them use loglevel 0?

Have you run the code?
If take both 2 and 3, a lot of log would be meaningless. 
Since the filter is valid, the enabled functions are what you configured.
What is the meaning of taking both 2 and 3?
It doesn't make sense to me.